proxylib/proxylib/connection.go

```diff
@@ -59,6 +62,117 @@ type Connection struct {
 	ReplyBuf   InjectBuf // Buffer for injected frames in reply direction
 }
 
+func NewConnection(instance *Instance, proto string, connectionId uint64, ingress bool, srcId, dstId uint32, srcAddr, dstAddr, policyName string, origBuf, replyBuf *[]byte) (FilterResult, *Connection) {
+	// Find the parser for the proto
+	parserFactory := GetParserFactory(proto)
+	if parserFactory == nil {
+		return UNKNOWN_PARSER, nil
+	}
+	_, port, err := net.SplitHostPort(dstAddr)
+	if err != nil {
+		return INVALID_ADDRESS, nil
+	}
+	dstPort, err := strconv.ParseUint(port, 10, 32)
+	if err != nil || dstPort == 0 {
+		return INVALID_ADDRESS, nil
+	}
+
+	connection := &Connection{
+		Instance:   instance,
+		Id:         connectionId,
+		Ingress:    ingress,
+		SrcId:      srcId,
+		DstId:      dstId,
+		SrcAddr:    srcAddr,
+		DstAddr:    dstAddr,
+		Port:       uint32(dstPort),
+		PolicyName: policyName,
+		ParserName: proto,
+		OrigBuf:    origBuf,
+		ReplyBuf:   replyBuf,
+	}
+	connection.Parser = parserFactory.Create(connection)
+	if connection.Parser == nil {
+		// Parser rejected the new connection based on the connection metadata
+		return POLICY_DROP, nil
+	}
+
+	return OK, connection
+}
+
+// Skip bytes in input, or exhaust the input.
+func advanceInput(input [][]byte, bytes int) [][]byte {
+	for bytes > 0 && len(input) > 0 {
+		rem := len(input[0]) // this much data left in the first slice
+		if bytes < rem {
+			input[0] = input[0][bytes:] // skip 'bytes' bytes
+			bytes = 0
+		} else { // go to the beginning of the next unit
+			bytes -= rem
+			input = input[1:] // may result in an empty slice
+		}
+	}
+	return input
+}
+
+func (connection *Connection) OnData(reply, endStream bool, data *[][]byte, filterOps *[][2]int64) (res FilterResult) {
```

Comment:
I did have a `C.FilterOp` type there, but hit difficulty due to the same `types.h` file included from `main` and `proxylib` packages defining their own definitions of that type. I have not fould examples of using C types of another package, but maybe it would be possible to include `types.h` only in the `proxylib/proxylib` package and then use `proxylib.C.FilterOp` etc. here instead?

Reply:
Hmm, yeah, you are right. I think the `C` is local to the package the import happens in, despite it being uppercase. My only concern was getting the indexes mixed. A comment somewhere is probably good enough (I guess you could also make named variables for each index but that feels a little yucky).